### ISSUE TITLE
How do I send a tweet with an emoji?

### ISSUE THREAD

**Issue by patrice-truong** [Yortw/tweetmoasharp#56]
Haven't found how to do so.

The Can_parse_multibyte_entities() unit test definitely shows that tweets with emoji can be read :), but how do I send one with an emoji in it? 

Using the official Twitter API (https://github.com/twitter/twemoji), for example, I can replace the thumb up emoji with <img class="emoji" draggable="false" alt="👍"  src="https://twemoji.maxcdn.com/2/72x72/1f44d.png">. I just haven't found how to send this html fragment with the tweetmoasharp library. Or is there an easier way?

Thanks!

**Comment by Yortw** [comment]
Hi,

Sorry for the delayed response, I have been away and was not in a position to do development/answer questions.

> How do I send a tweet with an emoji?

You include the unicode characters for the emoji in the string that has the tweet content. You can do this using any method that works in .Net for placing a unicode character in a string. The most common ways I'm aware of are to literally include the character (either holding alt and typing unicode char number on the keypad, or using an app to pick the symbol from a list and insert it via clipboard), or using the \u escape sequence (https://msdn.microsoft.com/en-us/library/aa664669(v=vs.71).aspx). You can also copy unicode characters from web pages that use the actual unicode character instead of an image (e.g. http://unicodeemoticons.com/)

So

`
service.SendTweet(new SendTweetOptions { Status = "Hello ☺" });
`

or

`
service.SendTweet(new SendTweetOptions { Status = "Hello \u263A" });
`

Should both send a tweet with a smiley emoji.

> Using the official Twitter API (https://github.com/twitter/twemoji), for example, I can replace the thumb up emoji with . 

I was not previously aware of this API, but with a quick half hour glance at the docs and related blog posts, it appears to be a library for *displaying* tweets *in a web platform*. It has nothing to do with sending tweets or putting specific content in a tweet as far as I can tell. It requires a script tag in the HTML header (doesn't work outside of web pages, tweets are often displayed outside of the web) and it's two main methods (overloads of 'parse') are about finding emoji unicode characters in the tweet and replacing the native characters with HTML image tags (again, only works in web/HTML platforms, tweets are not web pages).

>  I just haven't found how to send this html fragment with the tweetmoasharp library. 

AFAIK from my experience both as a Twitter user and 3rd party developer, you cannot include arbitary HTML inside a tweet. That would cause all sorts of problems, from security to user experience. Including HTML inside a tweet should sanitize the input such that the raw HTML is displayed, rather using that to affect display of the tweet content. Again, looking at te twemoji library you linked to, it doesn't seem to be used for changing the content of a tweet before it is posted, rather it seems to be about modifying the content of a tweet retrieved from Twitter for display in a web browser. I believe you should send the normal unicode characters for emoji, and whatever is displaying the tweet may or may not choose to use twemoji to display it. 

Good luck!

**Comment by patrice-truong** [comment]
Thanks for getting back to me on this issue..

I must be doing something wrong because  service.SendTweet(new SendTweetOptions { Status = "Hello \u263A" }); does not send an emoji. it sends the content verbatim, i.e. "Hello \u263A".

But if it works for you, then there must be something else in my application that causes the Unicode character not to be recognized. 

And yes, you were right for the twemoji API. It is just for displaying emoji.

**Comment by Yortw** [comment]
It does work me, not sure why it wouldn't work for you. If you are running in the debugger and put a watch on the string, it should show the emoji there too. For example, if I place a watch ony SendTweetOptions instance I see this;

![image](https://cloud.githubusercontent.com/assets/6725891/22264318/58ac289e-e2dc-11e6-8519-ad70341a4888.png)

Then when I actually send the tweet, I see this on Twitter;

![image](https://cloud.githubusercontent.com/assets/6725891/22264341/6ab66766-e2dc-11e6-9f8a-3c51fbb83047.png)

Does it work for you if you embed the literal unicode character instead of the escape sequence?

**Comment by Yortw** [comment]
That was literally sent using this code, where consumerKey, consumerSecret, userKey, and userSecret are strings containing the relevant credentials.

`
var service = new TweetSharp.TwitterService(consumerKey, consumerSecret, userKey, userSecret);
service.SendTweet(new SendTweetOptions { Status = "Hello \u263A" });
`

**Comment by patrice-truong** [comment]
OK. I definitely get the emoji in a test application. I just need to get it to work with my real application :)

Thanks so much for taking the time to work on my issue. Really appreciate!

**Comment by Yortw** [comment]
No problem, I hope your get it working. Come back if you find a problem or have another question.
Good luck!

**Comment by patrice-truong** [comment]
Sorry for the delayed response. I've been caught on other things..

I just did some more testing and it seems to only work with a few emojis. I am using emojis from https://www.piliapp.com/twitter-symbols/. I had a look at the source page (to get the HEX codes) and it seems that the ones that work with Tweetmoasharp are the ones with a HEX code < 10000.

[var service = new TweetSharp.TwitterService(consumerKey, consumerSecret, userKey, userSecret); service.SendTweet(new SendTweetOptions { Status = "Hello 😀☺️" });]

does not seem to work
Any idea why (and how to fix it?)

**Comment by patrice-truong** [comment]
![image](https://cloud.githubusercontent.com/assets/11589302/22625249/4d199cce-eb91-11e6-8086-fe8acb2d133e.png)
and btw, here's how my status looks like in the debugger.

**Comment by Yortw** [comment]
Hi,

You are correct, it doesn't work. Unfortunately I do not know why. There is a method inside the Hammock library which encodes characters like <space>, / + etc. in the url, and I suspect it is encoding each byte rather than the surrogate pair. However even if I remove this method it doesn't work, though the text contains the correct characters as far as I can tell up until it is sent via the .Net HttpRequest object. At this stage I'm still looking for a solution. PR's welcome.

**Comment by collinsauve** [comment]
Oh Unicode, how I loathe thee!  

Just to make our lives more difficult, I'm actually going to take a more complex example: 🙏🏽
This is an emoji succeeded by a surrogate pair that changes the color:  🙏 + 🏽 = 🙏🏽


The problem is in Hammock's [OAuthTools.UrlEncodeStrict](https://github.com/Yortw/hammock/blob/master/src/net35/Hammock/Authentication/OAuth/OAuthTools.cs).  It does not seem to properly encode multi-char characters.  The core problem is that it gets the char array, then gets bytes for each char individually.  You need to get bytes for the entire string or char array at once or else you will get the wrong bytes.


Here's a demonstration of this principal in action:

```
var message = "🙏🏽"; // Surprisingly VisualStudio and C# seem to be able to handle this as a literal
var chars = message.ToCharArray();
Console.WriteLine("Comma separator represents each supposed char");
Console.WriteLine("chars: " + string.Join(",", chars.Select(c => string.Format($"{(uint)c:X}"))));
var bytes = Encoding.UTF8.GetBytes(message);
Console.WriteLine("bytes: " + string.Join(",", chars.Select(b => string.Format($"{(uint)b:X}"))));
var charsThenBytes = message.ToCharArray().Select(c => Encoding.UTF8.GetBytes(new[] {c}));
Console.WriteLine("charsThenBytes: " + string.Join(",", charsThenBytes.Select(bytesForChar => string.Join("", bytesForChar.Select(b => string.Format($"{(uint)b:X}"))))));
```

Output:

```
Comma separator represents each supposed char
chars: D83D,DE4F,D83C,DFFD
bytes: D83D,DE4F,D83C,DFFD
charsThenBytes: EFBFBD,EFBFBD,EFBFBD,EFBFBD

```

Notice that what you go to chars and then bytes then each char becomes the bytes EFBFBD - each char and no matter which emoji this is.

Fortuntately, I have a fix for this.  Unfortunately I'm not using TweetMoaSharp anymore (mostly due to disliking Hammock).  I'll post this here and let you try to adapt it to TweetMoaSharp.  You can see my fix in action here: https://twitter.com/collinsauve/status/916027271437328385

Here's what I did:

```
private static readonly ISet<byte> UnreservedBytes = new HashSet<byte>(Encoding.UTF8.GetBytes(Unreserved.ToCharArray()));

public static string UrlEncodeStrict(string s)
{
    var result = new StringBuilder();
    var bytes = Encoding.UTF8.GetBytes(s);
    foreach (var b in bytes)
    {
        if (UnreservedBytes.Contains(b))
        {
            result.Append(Encoding.UTF8.GetString(new [] { b }));
        }
        else
        {
            // Support proper encoding of special characters (\n\r\t\b)
            if ((b > 7 && b < 11) || b == 13)
            {
                result.Append($"%0{b:X}");
            }
            else
            {
                result.Append($"%{b:X}");
            }
        }
    }
    return result.ToString();
}
```

**Comment by collinsauve** [comment]
I was not happy with the number of heap allocations with the above algorithm - especially for unreserved characters.  Note that in the above algo that even regular characters have to go through this for every character: `Encoding.UTF8.GetString(new [] { b }`.

So first I wrote some tests around my current algorithm:

```
[TestCase("Test", "Test", TestName = "No encoding")]
[TestCase("🙏", "%F0%9F%99%8F", TestName = "Emoji")]
[TestCase("🙏🏽", "%F0%9F%99%8F%F0%9F%8F%BD", TestName = "Surrogate Pair")]
[TestCase("()/%", "%28%29%2F%25", TestName = "Symbols")]
[TestCase("Nⓔｗ ̵̸̲̲̄Ｎ̵̸̲̲̄eｗ ϒ̶̷̲̲߫0ʀ̵̸̲̲̄ｋ ̵̸̲̲߫Tⓘｍᴇ̵̷̲̲̅ｓ", "N%E2%93%94%EF%BD%97%20%CC%B5%CC%B8%CC%B2%CC%B2%CC%84%EF%BC%AE%CC%B5%CC%B8%CC%B2%CC%B2%CC%84e%EF%BD%97%20%CF%92%CC%B6%CC%B7%CC%B2%CC%B2%DF%AB0%CA%80%CC%B5%CC%B8%CC%B2%CC%B2%CC%84%EF%BD%8B%E2%80%85%CC%B5%CC%B8%CC%B2%CC%B2%DF%ABT%E2%93%98%EF%BD%8D%E1%B4%87%CC%B5%CC%B7%CC%B2%CC%B2%CC%85%EF%BD%93", TestName = "Combining Diacritical Marks")]
[Test]
public void UrlEncodeStrict(string input, string expectedOutput)
{
    var output = OAuthEncoding.UrlEncodeStrict(input);
    Assert.AreEqual(expectedOutput, output);
}
```

Then I rewrote it.  Turns out you can do `UTF8.GetBytes` on surrogate pairs from a char array, as long as locate them and then pass them together:

```
private static readonly ISet<char> UnreservedChars = new HashSet<char>(Unreserved.ToCharArray());

public static string PercentEncode(this string s) => new string(PercentEncode(s.ToCharArray()).ToArray());

private static IEnumerable<char> PercentEncode(char[] chars)
{
    if (chars == null)
    {
        yield break;
    }

    var i = 0;
    while (i < chars.Length)
    {
        var thisChar = chars[i];
        if (UnreservedChars.Contains(thisChar))
        {
            i += 1;
            yield return thisChar;
        }
        else
        {
            var nChars = GetNumberOfCharsForCharacter(chars, i);
            var bytes = Encoding.UTF8.GetBytes(chars, i, nChars);
            foreach (var b in bytes)
            {
                // Support proper encoding of special characters (\n\r\t\b)
                var resultString = (b > 7 && b < 11) || b == 13
                    ? $"%0{b:X}"
                    : $"%{b:X}";
                foreach (var resultChar in resultString.ToCharArray())
                {
                    yield return resultChar;
                }
            }
            i += nChars;
        }
    }
}

public static int GetNumberOfCharsForCharacter(char[] str, int index)
{
    var firstChar = str[index];
    var isSurrogate = char.GetUnicodeCategory(firstChar) == UnicodeCategory.Surrogate;
    if (!isSurrogate)
    {
        return 1;
    }
    if (index + 1 >= str.Length)
    {
        throw new ArgumentException($"Character at position {index} is a Surrogate but is the last character in the string.");
    }
    if (char.GetUnicodeCategory(str[index + 1]) != UnicodeCategory.Surrogate)
    {
        throw new ArgumentException($"Character at position {index} is a Surrogate but the next character is not a Surrogate.");
    }
    return 2;
}
```

**Comment by Yortw** [comment]
Should be fixed in 4.0.3 which was just pushed to Nuget. Thanks to @collinsauve
